internal/fields/validate.go

```diff
@@ -103,6 +103,9 @@ func (v *Validator) validateMapElement(root string, elem common.MapStr) multierr
 				}
 			}
 		case map[string]interface{}:
+			if definition := findElementDefinition("", key, v.schema); definition != nil && definition.Type == "flattened" {
```

Comment:
I see that you reused the same invocation as below (`findElementDefinition("", key, v.schema)`). I think it's a good moment to refactor it a bit:

`findElementDefinition(key, v.schema)` which calls `findElementDefinitionForRoot("", key, v.schema)`

and apply then same in `validateScalarElement()`


I would also wrap the entire line with `IsFieldFlattened` or similar.